add isFileSystemDate flag to thesaurus model

wangf1122 opened the pull request:
This is to fix the issue https://github.com/metadata101/iso19139.ca.HNAP/issues/308

The issue is the schema transformation style sheet doesn't know if the date is from the thesaurus itself or from the file system last modified date. Therefore, it cannot adjust logic to render different XML template.

ianwallen commented:
@wangf1122 

I don't believe this is the fix that we are looking for....

@josegar74 @fxprunayre 

I think the real bug in this case is that it should never set the last modified date to the system date or the file date. 

In our case, we are regularly doing deployments in an K8S environment, and we always copy the thesaurus files during the deployment - this does not mean that the last modified date for the thesaurus was changed.
With this logic, the thesaurus dates on our application are changing almost every time a user does an update.

I believe it should instead just leave it last modified date empty and not modify the values that already exists in the metadata files.  If someone really wants a last modified date set then the date should be managed in the thesaurus file.

What are your thoughts?

josegar74 commented:
@ianwallen the "problem" is that https://github.com/geonetwork/core-geonetwork/blob/main/schemas/iso19139/src/main/plugin/iso19139/convert/thesaurus-transformation.xsl creates all the thesaurus block from scratch, probably as only receives the parameters of the keyword(s) to add and not the full metadata record.

For me the solution would be to store the creation/publish dates in the vocabulary file, which is possible. The current code  server-side code retrieves the modification date (if it exists). 

https://github.com/geonetwork/core-geonetwork/blob/d7a4b80c109da43a2b7a747c69f7c7fd481bdf3e/core/src/main/java/org/fao/geonet/kernel/Thesaurus.java#L968-L980

I think it can also be updated to handle the creation date.

wangf1122 commented:
> @ianwallen the "problem" is that https://github.com/geonetwork/core-geonetwork/blob/main/schemas/iso19139/src/main/plugin/iso19139/convert/thesaurus-transformation.xsl creates all the thesaurus block from scratch, probably as only receives the parameters of the keyword(s) to add and not the full metadata record.
> 
> For me the solution would be to store the creation/publish dates in the vocabulary file, which is possible. The current code server-side code retrieves the modification date (if it exists).
> 
> https://github.com/geonetwork/core-geonetwork/blob/d7a4b80c109da43a2b7a747c69f7c7fd481bdf3e/core/src/main/java/org/fao/geonet/kernel/Thesaurus.java#L968-L980
> 
> I think it can also be updated to handle the creation date.

I have modified this pull request to pass all possible dates (created, modified, issued) to the schema level. And will let schema decides what to do with the dates. The original date logic is still there which I see it first uses issued date, then go modified. If none is found, then it will use the file system date.

I will make some PR at HNAP metadata to handle this logic. So this change will also give all schemas more flexibilities to deal with these dates.

ianwallen commented on core/src/main/java/org/fao/geonet/kernel/Thesaurus.java:
```diff
@@ -965,18 +983,29 @@ private void retrieveThesaurusInformation(Path thesaurusFile, String defaultTitl
 
-            Element dateEl = Xml.selectElement(thesaurusEl, "skos:ConceptScheme/dcterms:issued|skos:Collection/dc:date", theNSs);
+            Element issuedDateEl = Xml.selectElement(thesaurusEl, "skos:ConceptScheme/dcterms:issued|skos:Collection/dc:date", theNSs);
+            DateFormat dateFormatYearOnly = new SimpleDateFormat("yyyy");
```
Why is the format YYYY? It could be in many different formats we don't know what format it will be in.
Why are we  checking any format?  Why not simply take the value from dcterms:issued exacly as entered as a text string?

wangf1122 replied:
I have the new code to use the text as is.

ianwallen commented on core/src/main/java/org/fao/geonet/kernel/Thesaurus.java:
```diff
@@ -965,13 +983,20 @@ private void retrieveThesaurusInformation(Path thesaurusFile, String defaultTitl
             }
 
-            Element dateEl = Xml.selectElement(thesaurusEl, "skos:ConceptScheme/dcterms:issued|skos:Collection/dc:date", theNSs);
+            Element issuedDateEl = Xml.selectElement(thesaurusEl, "skos:ConceptScheme/dcterms:issued|skos:Collection/dc:date", theNSs);
```
I think issuedDateEl should only be `skos:ConceptScheme/dcterms:issued`

dateEl should be the old value `skos:ConceptScheme/dcterms:issued|skos:Collection/dc:date` or it should be set to issuedDateEl  and if null then attempt to use `skos:Collection/dc:date`

wangf1122 replied:
i had reverted the default date code to exactly original. So the only change are the newly added created, issued, modified date

ianwallen replied:
You still did not remove dc:date from issuedDateEI

wangf1122 replied:
removed

ianwallen commented on core/src/main/java/org/fao/geonet/kernel/Thesaurus.java:
```diff
@@ -965,8 +983,17 @@ private void retrieveThesaurusInformation(Path thesaurusFile, String defaultTitl
+
+            // Default date
+            Element dateEl = Xml.selectElement(thesaurusEl, "skos:ConceptScheme/dcterms:issued|skos:Collection/dc:date", theNSs);
             Date thesaususDate = parseThesaurusDate(dateEl);
```
For HNAP core subject thesaurus, the issued data would be 2004 which is an acceptable dc date.

However the parseThesaurusDate generates the following error message.

```
2023-04-20 23:18:34,908 ERROR [geonetwork.thesaurus-man] - Error parsing the thesaurus date value: 2004
2023-04-20T23:18:34.922650900Z   * with format: EEE MMM d HH:mm:ss z yyyy. Error is: Unparseable date: "2004"
2023-04-20T23:18:34.922661300Z   * with format: yyyy-MM-dd HH:mm:ss. Error is: Unparseable date: "2004"
2023-04-20T23:18:34.922667200Z   * with format: yyyy-MM-dd. Error is: Unparseable date: "2004"
2023-04-20T23:18:34.922673600Z Check thesaurus date in GC_Core_Subject.rdf
```

And since it cannot determine the date it will use a system date. 

In this case I think we should add the following below in the parseThesaurusDate so that it will end up using jan 1st 2004.

dfList.add(new SimpleDateFormat("yyyy"));

The following is also acceptable if we don't know the day of the month. So this could be added as well.
dfList.add(new SimpleDateFormat("yyyy-MM"));

wangf1122 commented:
@josegar74 

I have cherry picked your commit into this pull request. All good. Thanks.

fxprunayre commented on schemas/iso19115-3.2018/src/main/plugin/iso19115-3.2018/convert/thesaurus-transformation.xsl:
```diff
@@ -299,6 +305,87 @@
+                </cit:CI_Date>
+              </cit:date>
+            </xsl:when>
+            <xsl:otherwise>
```
One issue here is that if the thesaurus has no date (eg. ontologies from EU DCAT profiles https://mobilitydcat-ap.github.io/controlled-vocabularies/mobility-theme/latest/index.html), then the date is added twice. I'll look for a fix.

fxprunayre replied:
Improved in https://github.com/geonetwork/core-geonetwork/pull/8392
